btree/btree.c

```diff
@@ -2377,6 +2411,288 @@ void m0_btree_ut_node_add_del_rec(void)
+
+	m0_btree_open(temp_node, 1024, &btree);
+
+	m0_btree_close(btree);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	m0_free_aligned(temp_node, (1024 + sizeof(struct nd)), 10);
+
+	/** Now run some invalid cases */
+
+	/** Open a non-existent btree */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Close a non-existent btree */
+	m0_btree_close(btree);
+
+	/** Destroy a non-existent btree */
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	/** Create a new btree */
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op,
+			      m0_btree_create(temp_node, 1024, &btree_type,
+					      tx, &b_op));
+
+	/** Close it */
+	m0_btree_close(b_op.bo_arbor);
+
+	/** Try closing again */
+	m0_btree_close(b_op.bo_arbor);
+
+	/** Re-open it */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Open it again */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Destory it */
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	/** Attempt to reopen the destroyed tree */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+}
+
+
+struct cb_data {
+	struct m0_btree_key *key;
+	struct m0_bufvec    *value;
+};
+
+static int btree_kv_put_cb(struct m0_btree_cb *cb, struct m0_btree_rec *rec)
+{
+        struct m0_bufvec_cursor  scur;
+	struct m0_bufvec_cursor  dcur;
+	m0_bcount_t              ksize;
+	m0_bcount_t              vsize;
+	struct cb_data          *datum = cb->c_datum;
+
+	ksize = m0_vec_count(&datum->key->k_data.ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_key.k_data.ov_vec) >= ksize);
+
+	vsize = m0_vec_count(&datum->value->ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_val.ov_vec) >= vsize);
+
+	m0_bufvec_cursor_init(&scur, &datum->key->k_data);
+	m0_bufvec_cursor_init(&dcur, &rec->r_key.k_data);
+	m0_bufvec_cursor_copy(&dcur, &scur, ksize);
+
+	m0_bufvec_cursor_init(&scur, datum->value);
+	m0_bufvec_cursor_init(&dcur, &rec->r_val);
+	m0_bufvec_cursor_copy(&dcur, &scur, vsize);
+
+	return 0;
+}
+
+
+static int btree_kv_get_cb(struct m0_btree_cb *cb, struct m0_btree_rec *rec)
+{
+	struct m0_bufvec_cursor scur;
+	struct m0_bufvec_cursor dcur;
+	m0_bcount_t             ksize;
+	m0_bcount_t             vsize;
+	struct cb_data          *datum = cb->c_datum;
+
+	ksize = m0_vec_count(&datum->key->k_data.ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_key.k_data.ov_vec) <= ksize);
+
+	vsize = m0_vec_count(&datum->value->ov_vec);
+	M0_PRE(m0_vec_count(&rec->r_val.ov_vec) <= vsize);
+
+	m0_bufvec_cursor_init(&dcur, &datum->key->k_data);
+	m0_bufvec_cursor_init(&scur, &rec->r_key.k_data);
+	m0_bufvec_cursor_copy(&dcur, &scur, ksize);
+
+	m0_bufvec_cursor_init(&dcur, datum->value);
+	m0_bufvec_cursor_init(&scur, &rec->r_val);
+	m0_bufvec_cursor_copy(&dcur, &scur, vsize);
+
+	return 0;
+}
+
+
+/**
+ * This unit test exercises the KV operations for both valid and invalid
+ * conditions.
+ */
+static void m0_btree_ut_basic_kv_operations(void)
```

Comment:
`basic` can be removed from function name.

Reply:
I added it because there would be some more ut's which would exercise the same function calls but with more parallelism.